crates/p4-analyzer/src/driver.rs

```diff
@@ -0,0 +1,243 @@
+use analyzer_host::{json_rpc::message::Message, MessageChannel};
+use async_channel::{Receiver, Sender};
+#[cfg(test)]
+use async_std::task::block_on;
+use cancellation::{CancellationToken, OperationCanceled};
+use std::{
+	io::{stdin, stdout},
+	sync::{Arc},
+};
+use tokio::task;
+
+/// Connects the `stdin` and `stdout` of the process to appropriate [`MessageChannel`] instances, and executes a sender and
+/// reader thread to marshal [`Message`] instances between them.
+pub struct Driver {
+	in_channel: MessageChannel, 	// Input channel as an intermediate layer for input source -> Analyzer host
+	out_channel: MessageChannel, 	// Output channel as an intermediate layer for Analyzer host -> output source
+	source_type: DriverType,    	// Handles the connection for input/output of outside layer to intermediate layer
+}
+
+#[derive(Clone)]
+pub enum DriverType {
+	Console,
+	#[cfg(test)]
+	Buffer(BufferStruct), // Only want to include this for testing
+}
+
+pub fn console_driver() -> Driver {
+	Driver {
+		in_channel: async_channel::unbounded::<Message>(),
+		out_channel: async_channel::unbounded::<Message>(),
+		source_type: DriverType::Console,
+	}
+}
+
+impl Driver {
+	pub fn new(driver_type: DriverType) -> Driver {
+		match driver_type {
+			DriverType::Console => console_driver(),
+			#[cfg(test)]
+			DriverType::Buffer(buffer) => buffer_driver(buffer),
+		}
+	}
+
+	/// Retrieves a [`MessageChannel`] from which [`Message`] instances can be received from (i.e., `stdin`) and sent to (i.e., `stdout`).
+	pub fn get_message_channel(&self) -> MessageChannel {
+		let (sender, _) = self.out_channel.clone();
+		let (_, receiver) = self.in_channel.clone();
+
+		(sender, receiver)
+	}
+
+	fn sender_task(sender: Sender<Message>, input_source: DriverType) {
+		let match_func = || match input_source.clone() {
+			DriverType::Console => Message::read(&mut stdin().lock()),
+			#[cfg(test)]
+			DriverType::Buffer(mut buffer) => block_on(buffer.message_read()),
+		};
```

Comment:
Perhaps this could be added to an `impl` block for `DriverType` instead? It would hide the underlying implementation.